aiomisc/log/formatter/json.py

```diff
@@ -0,0 +1,101 @@
+import json
+import logging
+import sys
+import traceback
+from types import MappingProxyType
+from typing import Union
+
+
+class JSONLogFormatter(logging.Formatter):
+    LEVELS = MappingProxyType({
+        logging.CRITICAL: "crit",
+        logging.FATAL: "fatal",
+        logging.ERROR: "error",
+        logging.WARNING: "warn",
+        logging.WARN: "warn",
+        logging.INFO: "info",
+        logging.DEBUG: "debug",
+        logging.NOTSET: None,
+    })
+
+    FIELD_MAPPING = MappingProxyType({
+        'filename': ('code_file', str),
+        'funcName': ('code_func', str),
+        'lineno': ('code_line', int),
+        'module': ('code_module', str),
+        'name': ('identifier', str),
+        'msg': ('message_raw', str),
+        'process': ('pid', int),
+        'processName': ('process_name', str),
+        'threadName': ('thread_name', str),
+    })
+
+    def __init__(self, fmt=None, datefmt=None, style='%'):
+        super().__init__(
+            datefmt=datefmt if datefmt is not Ellipsis else None,
+            fmt=fmt, style=style
+        )
+
+    def format(self, record: logging.LogRecord):
+        record_dict = MappingProxyType(record.__dict__)
+
+        data = dict(errno=0 if not record.exc_info else 255)
+
+        for key, value in self.FIELD_MAPPING.items():
+            mapping, field_type = value
+
+            v = record_dict.get(key)
+
+            if not isinstance(v, field_type):
+                v = field_type(v)
+
+            data[mapping] = v
+
+        for key in record_dict:
+            if key in data:
+                continue
+            elif key[0] == "_":
+                continue
+
+            value = record_dict[key]
+
+            if value is None:
+                continue
+
+            data[key] = value
+
+        for idx, item in enumerate(data.pop('args', [])):
+            data['argument_%d' % idx] = str(item)
+
+        payload = {
+            '@fields': data,
+            'msg': self.formatMessage(record),
+            'level': self.LEVELS[record.levelno],
+        }
+
+        if self.datefmt:
+            payload['@timestamp'] = self.formatTime(record, self.datefmt)
+
+        if record.exc_info:
+            payload['stackTrace'] = self.formatException(record.exc_info)
+
+        return json.dumps(payload, ensure_ascii=False)
```

Comment:
maybe we should add support for custom json serializer

Reply:
ok.